### PR TITLE
AddLiquidity Custom Hooks

### PR DESCRIPTION
**Pull request by Pochetes**
`NOTE` there are 14 commits for this PR because the `develop` branch was based on `main` which is super behind from branches like `Godwin` which I based my branch on. Might get conflicts when merging `Godwin` into `develop`.

* was able to create a custom hook for `AddLiquidity` which dramatically decreased LOC for the component and helps for debugging the logic.
* Still stuck on `validateSubmit` function that conditionally returns a component whenever state like token input or token selection changes - still works by leaving it on UI component but want to figure out how to move it to custom hook while keeping functionality.

Any pointers on how to do this?

### CODE REVIEW

**Comment by TBomberman** [comment]
buttonStyle and buttonText can be set to a state variable which will change the button whenever the variable changes. Would that help?

File: pages/positions/addliquidity.tsx [hunk 1]
```diff
@@ -22,51 +13,33 @@ const style = {
...
-    setToken(tokenInput.replace(/[^0-9.]/g, ''))
-  }
-
-  // checks which tokenSelector element was selected and opens modal
-  const handleTokenSelector = (tokenSelected: string): void => {
-    setTokenSelected(tokenSelected)
-    setIsOpen(true)
-  }
+  const { 
+    tokenAInputVal,
```
Comment (TBomberman) [inline pages/positions/addliquidity.tsx:28]: So it looks like all the values and handlers are in the handlers file as another way of being a controller. Is there a name for this architectural style with guidance and guidelines to follow?

File: utils/getSmartContract.tsx [hunk 2]
```diff
@@ -1,39 +1,108 @@
...
+    provider: Provider
+): Promise<AmountsOut | undefined> => {
+    const routerContract = new ethers.Contract(process.env.IUNISWAP_V2_ROUTER_02_ADDR as string, IUniswapV2Router02ABI, provider as Provider)
+    let convertedInput = ethers.utils.parseEther(inputAmt)
+    const amountsOut = await routerContract.getAmountsOut(convertedInput, [tokenAddrs[0], tokenAddrs[1]])
+    return amountsOut
+}
+
+
+// /**
```
Comment (TBomberman) [inline utils/getSmartContract.tsx:54]: should remove commented code

File: utils/getSmartContract.tsx [hunk 3]
```diff
@@ -1,39 +1,108 @@
-import { ethers } from 'ethers'
+import { useContext } from 'react'
```
Comment (TBomberman) [inline utils/getSmartContract.tsx:1]: I'm not sure what is the purpose of this file. Is it to get all contracts related to a pair? ie, uniswap pair, uni router, uni factory, tokens, gs factory, gs pool, gs pos mgr? If so, is it updating all revelavent contracts on token update?

Comment (Pochetes) [reply utils/getSmartContract.tsx:1]: This was supposed to be a test file where I tried connecting smart contracts and testing out their functions. I realized that this was confusing to you all so I'll be removing it soon. In turn, I'll propose something that can conveniently solve this

File: utils/getSmartContract.tsx (hunk 2 above)
Comment (Pochetes) [reply utils/getSmartContract.tsx:54]: done, we will modify this file soon

File: components/Layout.tsx [hunk 4]
```diff
@@ -12,11 +13,13 @@ const style = {
 const Layout = ({ children }: Props) => {
     return (
         <WalletProvider>
-            <Head>
-                <title>GammaSwap</title>
-                <link rel="icon" href="/favicon.ico" />
-            </Head>
-            <main className={style.wrapper}>{children}</main>
+            <BasicContractProvider>
```
[the remark this answers is not in the document]
Comment (Pochetes) [reply components/Layout.tsx:16]: Definitely. Mainly to prevent [callback hell](https://www.reddit.com/r/javascript/comments/5rjrcy/why_is_twospace_tab_indent_becoming_the_standard/). When we get his changes merged, it'll adjust automatically.

File: context/BasicContractContext.tsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    IUniswapV2PairContract: Contract | null
+    IUniswapV2FactoryContract: Contract | null
+    IUniswapV2Router02Contract: Contract | null
+}
+
+type BasicContractProviderProps = {
+    children: ReactNode
+}
+
+export const BasicContractContext = createContext<BasicContractType>({} as BasicContractType)
```
[the remark this answers is not in the document]
Comment (Pochetes) [reply context/BasicContractContext.tsx:22]: This was originally to globalize the smart contract instantiation. So that we wouldn't have to do
```javascript
const someSmartContract = new ethers.Contract(address, contractABI, provider)
```
every time we want to use a smart contract.

However, I realized the ethers library makes it so that you have to instantiate it on the same file. Which is weird. Will do more research on this when we have Daniel's new contracts and/or open a new issue to solve this on Ethers's repo

File: hooks/useAddLiquidityHandlers.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
+import {
+    useState,
```
[the remark this answers is not in the document]
Comment (Pochetes) [reply hooks/useAddLiquidityHandlers.ts:2]: Yeah, we need to build on top of Usama's prettier config and make our own eslint configuration. This helps us make less nitpick review on PRs. Which helps us move faster

[the remark this answers is not in the document]
Comment (Pochetes) [reply hooks/useAddLiquidityHandlers.ts:2]: WIll fix this in next PR after Usama's PR gets merged

File: hooks/useAddLiquidityHandlers.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+    // checks which tokenSelector element was selected and opens modal
+    const handleTokenSelector = (tokenSelected: string): void => {
+        setTokenSelected(tokenSelected)
+        setIsOpen(true)
+    }
+
+    // checks if token selected object is empty
+    const isTokenEmpty = (tokenToCheck: Token): boolean => {
+        return Object.values(tokenToCheck).every(tokenProp => tokenProp === "" || tokenProp === 18)
```
[the remark this answers is not in the document]
Comment (Pochetes) [reply hooks/useAddLiquidityHandlers.ts:72]: Yeah, we can make an exceptions list of all the tokens that fall into that category. It shouldn't be much to start

File: hooks/useAddLiquidityHandlers.ts [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    
+    // checks for non-numeric value inputs
+    const validateTokenInput = (
+        e: ChangeEvent<HTMLInputElement> | string,
+        setToken: Dispatch<SetStateAction<string>>
+    ): void => {
+        let tokenInput: string
+        if (typeof e !== "string") tokenInput = (e.target as HTMLInputElement).value
+        else tokenInput = e
+        setToken(tokenInput.replace(/[^0-9.]/g, ''))
```
[the remark this answers is not in the document]
Comment (Pochetes) [reply hooks/useAddLiquidityHandlers.ts:61]: It's still giving me a bug doing it this way. We also need to check for cases if `tokenInput` is prefixed with a period or a number between 0 and 1. These cases have been giving me errors

File: hooks/useAddLiquidityHandlers.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Pochetes) [reply hooks/useAddLiquidityHandlers.ts:72]: We should make this validation logic its own PR

[the remark this answers is not in the document]
Comment (Pochetes) [reply hooks/useAddLiquidityHandlers.ts:72]: For now, I'll merge as is

File: hooks/useAddLiquidityHandlers.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Pochetes) [reply hooks/useAddLiquidityHandlers.ts:61]: We should make this validation logic its own PR. For now, I'll merge as is

File: hooks/useAddLiquidityHandlers.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Pochetes) [reply hooks/useAddLiquidityHandlers.ts:72]: We'll add this in the next PR

File: pages/positions/addliquidity.tsx (hunk 1 above)
Comment (Pochetes) [reply pages/positions/addliquidity.tsx:28]: Yeah, this is React's form of the MVC architechture. Per [react's docs](https://reactjs.org/docs/hooks-custom.html), it's a `.ts` file that handles all of the logic that the normal component file would of had (all code before the `return`). Usually, React suggests to name these custom hook files prefixed with `use`, to identify it. 

The way you and Usama would do it from now on is like this. Since React does not reflect the MVC pattern (the view is already the component itself), you would not need to separate it as MVC. You can simply build the component (view) and custom hook as the logic (controller, and better if it can be reused in other components). What is great for us is that we don't need a Model, since the backend is decentralized and autonomous. 

**TL;DR** - guideline for creating any sort of component with logic is by having a normal component for UI and extracted logic into a custom hook prefixed by `use`

**Comment by Pochetes** [comment]
> buttonStyle and buttonText can be set to a state variable which will change the button whenever the variable changes. Would that help?

Yeah I tried doing that below
https://github.com/gammaswap/v1-interface/blob/9fa9968f4813284bae73052b64ba3675ace9a1f7/hooks/useAddLiquidityHandlers.ts#L47-L51

But it's giving me errors, I'm going to try it again in the next PR

File: pages/positions/addliquidity.tsx (hunk 1 above)
Comment (TBomberman) [reply pages/positions/addliquidity.tsx:28]: But this is for react state interaction I think. Is it suitable for async nature of web3?

Comment (TBomberman) [reply pages/positions/addliquidity.tsx:28]: Seems reasonable to have react hooks. Does this mean that all the hooks will be encoded in the same page? As the page is seems kind of bloated already.
